### PR TITLE
Add shape functions for conv_transpose2d.input and convolution

### PR DESCRIPTION
**Pull request by gpetters94**
As @silvasean requested in [this issue](https://github.com/llvm/torch-mlir/pull/917#discussion_r896154545) here is the shape code from Torch-MLIR for conv_transpose2d.input and convolution (updated for the transposed case).

### CODE REVIEW

**Comment by gpetters94** [comment]
> @gpetters94 i fired the tests but they're gonna fail (though I'm not sure which ones) because you haven't genereated the cpp definitions like here https://github.com/pytorch/pytorch/pull/79194/files. you need to run [gen_jit_shape_functions.py](https://github.com/pytorch/pytorch/blob/45cb8199df131a333615123e76c9cf5b37b7cdf1/torchgen/shape_functions/gen_jit_shape_functions.py#L14-L17).

Oh, I see. Doing that right now.

**Comment by gpetters94** [comment]
Sorry for all the small commits, still not used to locally testing PyTorch code.

**Comment by makslevental** [comment]
> Sorry for all the small commits, still not used to locally testing PyTorch code.

np but you're gonna have to squash before it can be merged

**Comment by gpetters94** [comment]
> > Sorry for all the small commits, still not used to locally testing PyTorch code.
> 
> np but you're gonna have to squash before it can be merged

Done, sorry about that.

**Comment by gpetters94** [comment]
@makslevental Okay, it's rebased to TOM and squashed, so it should be good to go.

**Comment by gpetters94** [comment]
@eellison Some of the tests are failing, but it doesn't look like the errors are due to my code. Do you think this can be merged as it is now, or should I be tracking these errors down?

**Review by Gamrix — CHANGES_REQUESTED** [review]
Could you add a assert_jit_shape_analysis=True in the relevant place like is done in the following diff so that we can test this function: https://github.com/pytorch/pytorch/commit/6b6c63ce5e0b54eb92cdd88f3bb87efc7adeaf9c

Also I recommend rebasing on viable/strict so that we can isolate you from irrelevant test failures.

**Comment by gpetters94** [comment]
> Could you add a assert_jit_shape_analysis=True in the relevant place like is done in the following diff so that we can test this function: [6b6c63c](https://github.com/pytorch/pytorch/commit/6b6c63ce5e0b54eb92cdd88f3bb87efc7adeaf9c)
> 
> Also I recommend rebasing on viable/strict so that we can isolate you from irrelevant test failures.

Good idea, I've done so.

File: torch/jit/_shape_functions.py
Comment (Gamrix) [inline torch/jit/_shape_functions.py:1035]:
```suggestion
add_shape_compute_mapping("aten::convolution(Tensor input, Tensor weight, Tensor? bias, int[] stride, int[] padding, int[] dilation, bool transposed, int[] output_padding, int groups) -> Tensor", conv_forwards)
add_shape_compute_mapping("aten::conv_transpose2d.input(Tensor input, Tensor weight, Tensor? bias=None, int[2] stride=1, int[2] padding=0, int[2] output_padding=0, int groups=1, int[2] dilation=1) -> Tensor)
```

The tests are erroring out due to the incorrect format.

**Review by Gamrix — CHANGES_REQUESTED** [review]
The tests are erroring out due to an incorrect shape compute mapping function schema name. I applied the changes and am rerunning tests.

**Comment by gpetters94** [comment]
> The tests are erroring out due to an incorrect shape compute mapping function schema name. I applied the changes and am rerunning tests.

Ah, I see. Thanks for the help!

**Comment by Gamrix** [comment]
I have recently pushed some fixes in this diff: https://github.com/pytorch/pytorch/pull/79571

I recommend rebasing on top of it. The easiest way to do so would be to rebase on top of viable/strict tomorrow (8/11) or later. Then ignore any merge conflict in the generated file and regenerate it.

**Comment by gpetters94** [comment]
> I have recently pushed some fixes in this diff: #83092
> 
> You will need to rebase on top of it. The easiest way to do so would be to rebase on top of viable/strict tomorrow (8/11) or later. Then ignore any merge conflict in the generated file and regenerate it.

Fantastic, thank you.

**Comment by Gamrix** [comment]
@albanD @seemethere This PR is incorrectly erroring out on the 2000 LOC maximum check in `[pr-sanity-checks](https://github.com/pytorch/pytorch/runs/7810904545?check_suite_focus=true#logs)` . There is only 124 LOC of changes. Could you check the logic for this code and see why this is happening?

**Comment by albanD** [comment]
@Gamrix sorry about that. The fix is here: https://github.com/pytorch/pytorch/pull/83344

File: torch/jit/_shape_functions.py
Comment (Gamrix) [inline torch/jit/_shape_functions.py:744]:
```suggestion

```

**Review by Gamrix — APPROVED** [review]
Looks good, now just waiting for all the tests to pass before I land this. 

(I will ignore the PR line count error)

**Comment by Gamrix** [comment]
Thanks @gpetters94 for contributing :D

**Comment by CaoE** [comment]
@gpetters94 May I know which tests can cover this change, and how do we use `conv_forwards` and `conv_transpose2d_input` defined in `serialized_shape_function_registry.cpp` or `_shape_functions.py`. 

Seems the code below can't utilize `conv_forwards`  since the node in the graph generated by torch.jit.trace for conv2d is `_convolution` instead of `convolution`, considering `{"aten::convolution(Tensor input, Tensor weight, Tensor? bias, int[] stride, int[] padding, int[] dilation, bool transposed, int[] output_padding, int groups) -> Tensor", "conv_forwards"}`
```
inp_shape = torch.rand([1, 16, 5, 10])
weight_shape = torch.rand([33, 16, 3, 3])
bias = None
stride = [2, 2]
padding = [0, 0]
dilation = [1, 1]
groups = 1
res = torch.nn.functional.conv2d(inp_shape, weight_shape, bias, stride, padding, dilation, groups).size()

def foo(inp_shape, weight_shape):
    return torch.nn.functional.conv2d(inp_shape, weight_shape, bias, stride, padding, dilation, groups)

fn = torch.jit.trace(foo, (inp_shape, weight_shape))
torch._C._jit_erase_non_input_shape_information(fn.graph)
torch._C._jit_pass_peephole(fn.graph)
torch._C._jit_pass_constant_propagation(fn.graph)
self.checkShapeAnalysis(res, fn.graph, assert_propagation=True)
```

**Pull request closed without merging**
